math/mathcore/inc/Math/CladDerivator.h

```diff
@@ -19,420 +19,221 @@
-  void Min_darg1_grad(T a, T b, T* result) {
-    return;
-  }
-
-  template <typename T>
-  T Power_darg0(T x, T y) {
-    return y * TMath::Power(x, y - 1);
-  }
-
-  template <typename T>
-  T Power_darg0_darg0(T x, T y) {
-    return y * (y - 1) * TMath::Power(x, y - 2);
-  }
-
-  template <typename T>
-  Double_t Power_darg0_darg1(T x, T y) {
-    return TMath::Power(x, y - 1) * (y * TMath::Log(x) + 1);
-  }
-
-  template <typename T>
-  Double_t Power_darg1(T x, T y) {
-    return TMath::Power(x, y) * TMath::Log(x);
-  }
-
-  template <typename T>
-  Double_t Power_darg1_darg0(T x, T y) {
-    return TMath::Power(x, y - 1) * (y * TMath::Log(x) + 1);
-  }
-
-  template <typename T>
-  Double_t Power_darg1_darg1(T x, T y) {
-    return TMath::Power(x, y) * TMath::Sq(TMath::Log(x));
-  }
-
-  template <typename T>
-  Double_t Power_grad(T x, T y, Double_t* result) {
-    T t = TMath::Power(x, y - 1);
-    result[0] += y * t;
-    result[1] += x * t * TMath::Log(x);
-  }
-
-  template <typename T>
-  Double_t Power_darg0_grad(T x, T y, Double_t* result) {
-    T t = TMath::Power(x, y - 2);
-    result[0] += y * (y - 1) * t;
-    result[1] += x * t * (y * TMath::Log(x) + 1);
-  }
-
-  template <typename T>
-  Double_t Power_darg1_grad(T x, T y, Double_t* result) {
-    T t = TMath::Power(x, y - 1);
-    result[0] += t * (y * TMath::Log(x) + 1);
-    result[1] += x * t * TMath::Sq(TMath::Log(x));
-  }
-
-  template <typename T>
-  Double_t Sin_darg0(T d) {
-    return TMath::Cos(d);
-  }
-
-  template <typename T>
-  Double_t Sin_darg0_darg0(T d) {
-    return -TMath::Sin(d);
-  }
-
-  template <typename T>
-  Double_t SinH_darg0(T d) {
-    return TMath::CosH(d);
-  }
-
-  template <typename T>
-  Double_t SinH_darg0_darg0(T d) {
-    return TMath::SinH(d);
-  }
-
-  template <typename T>
-  T Sq_darg0(T d) {
-    return 2 * d;
-  }
-
-  template <typename T>
-  T Sq_darg0_darg0(T d) {
-    return 2;
-  }
-
-  template <typename T>
-  Double_t Sqrt_darg0(T d) {
-    return 0.5 / TMath::Sqrt(d);
-  }
-
-  template <typename T>
-  Double_t Sqrt_darg0_darg0(T d) {
-    return 0.25 / (TMath::Sqrt(d) * d);
-  }
-
-  template <typename T>
-  Double_t Tan_darg0(T d) {
-    return 1./ TMath::Sq(TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t Tan_darg0_darg0(T d) {
-    return 2. * TMath::Sin(d) / (TMath::Cos(d) * TMath::Cos(d) * TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0(T d) {
-    return 1./ TMath::Sq(TMath::CosH(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0_darg0(T d) {
-    return -2. * TMath::SinH(d) / (TMath::CosH(d) * TMath::CosH(d) * TMath::CosH(d));
-  }
+namespace TMath {
+template <typename T>
+ValueAndPushforward<T, T> Abs_pushforward(T x, T d_x)
+{
+   return {::TMath::Abs(x), ((x < 0) ? -1 : 1) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<Double_t, Double_t> ACos_pushforward(T x, T d_x)
+{
+   return {::TMath::ACos(x), (-1. / ::TMath::Sqrt(1 - x * x)) * d_x};
```

Comment:
This will give an invalid result if `1 - x * x` is negative.

Reply:
Can you please check the derivatives for all related functions?

Reply:
@vgvassilev Since, the domain of ACos is `[-1, 1]`, should we give some error if the value `x` is outside this range?

Reply:
Let's ask for @lmoneta's preference.

Reply:
It is fine that we return a NaN if x is not in [-1,1]. This is also what is returned by `std::acos`. 

I would use here std functions and not TMath ones..

Reply:
> I would use here std functions and not TMath ones..

I agree with that, however, if there is a slight implementation differences in the `TMath` analogs to `std` then we might have troubles debugging. For example, if `TMath::Sin` and `std::sin` return slightly different results then the derivative of `TMath::Cos` (returning `std::sin` instead of `TMath::Sin`) would differ potentially making some sanity checks needing to compare approximate values. Not sure if that's a real problem for most of the functions.